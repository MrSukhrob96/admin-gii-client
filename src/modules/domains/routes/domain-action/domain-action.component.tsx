import { Button, Card, H1, H3 } from '@blueprintjs/core';
import { Error } from '@components/error';
import { InputField } from '@components/form/fields/input-field';
import { MultipleSelectField } from '@components/form/fields/multiple-select-field';
import { ArrayFieldCounter, ArrayFieldDelete } from '@styles/array-field';
import { FlexContainer, FlexItem } from '@styles/flex-container';
import { Space } from '@styles/space';
import { hoc } from '@utils/hoc';
import { useDomainActionProps } from './domain-action.props';
import {
  DomainActionContainer
} from './domain-action.style';

/**
 * <DomainAction />
 */

export const DomainAction = hoc(
  useDomainActionProps,
  ({ updatedItem, isCreate, onSubmit, roles, form, fieldsArrayField, fields }) => {
    if (!updatedItem && !isCreate)
      return <Error status='404' text='Bunday sahifa topilmadi' />;
    return (
      <DomainActionContainer>
        <H1>Yangi ma'lumot turi qo'shish</H1>
        <Space height='2rem' />
        <form className='w-half' onSubmit={form.handleSubmit(onSubmit)}>
          <InputField
            control={form.control}
            type='text'
            name='name'
            label='Nomi'
            rules={{ required: 'Nomini kiriting!' }}
          />
          <InputField
            control={form.control}
            type='text'
            name='slug'
            label='Slag'
            rules={{
              required: 'Slagni kiriting!',
              pattern: {
                value: /^[a-z0-9-_]+$/i,
                message:
                  'Slagda faqat lotin harflar, raqamlar, tire va pastki chiziq bo`ladi'
              }
            }}
          />
          <MultipleSelectField
            control={form.control}
            name='roles'
            label='Rollar'
            items={roles}
          />
          <Card>
            <FlexContainer align='center' wrap='wrap'>
              <FlexItem col={1}>
                <H3 className='bp4-heading'>Domain fields</H3>
              </FlexItem>
              {fieldsArrayField.fields.map((field, index) => (
                <FlexContainer key={field.id} gap='20px'>
                  <FlexItem col={50}>
                    <ArrayFieldCounter>
                      {index === 0 ? 'â„–' : index}
                    </ArrayFieldCounter>
                  </FlexItem>
                  <FlexItem col={2.12}>
                    <InputField
                      control={form.control}
                      type='text'
                      name={`fields.${index}.name`}
                      label='Fieldni nomi'
                      rules={{ required: 'Nomini kiriting!' }}
                    />
                  </FlexItem>
                  <FlexItem col={2.12}>
                    <InputField
                      control={form.control}
                      type='text'
                      name={`fields.${index}.slug`}
                      label='Fieldni slagi'
                      rules={{
                        required: 'Slagni kiriting!',
                        pattern: {
                          value: /^[a-z0-9-_]+$/i,
                          message:
                            'Slagda faqat lotin harflar, raqamlar, tire va pastki chiziq bo`ladi'
                        }
                      }}
                    />
                  </FlexItem>
                  {fieldsArrayField.fields.length > 1 && (
                    <FlexItem col={25}>
                      <ArrayFieldDelete>
                        <Button
                          icon={'trash'}
                          minimal={true}
                          onClick={() => fieldsArrayField.remove(index)}
                        />
                      </ArrayFieldDelete>
                    </FlexItem>
                  )}
                </FlexContainer>
              ))}
              <FlexItem col={1}>
                <Button
                  intent='primary'
                  icon='add'
                  onClick={() =>
                    fieldsArrayField.prepend({
                      name: '',
                      slug: ''
                    })
                  }
                >
                  Qo'shish
                </Button>
              </FlexItem>
            </FlexContainer>
          </Card>
          <Space height='1rem' />
          <MultipleSelectField
            control={form.control}
            name='table_fields'
            label="Tablitsada ko'rinadigan field-lar"
            items={fields}
          />
          <Button
            type='submit'
            intent='success'
            icon='cloud-upload'
            text='Saqlash'
          />
        </form>
      </DomainActionContainer>
    );
  }
);
