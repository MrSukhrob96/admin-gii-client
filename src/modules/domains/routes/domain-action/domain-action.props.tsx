import { OptionProps } from '@blueprintjs/core';
import { CreateDomainForm } from '@domains/model';
import { parseQueryToString } from '@utils/get-query';
import { useEffect, useMemo, useState } from 'react';
import { useFieldArray, useForm } from 'react-hook-form';
import { useHistory } from 'react-router-dom';

/**
 * <DomainAction /> props
 */
export type DomainActionProps = {};

const initialValues: CreateDomainForm = {
  name: '',
  slug: '',
  roles: [],
  fields: [{ name: '', slug: '', id: 'nx172eg2x61e71nxf2' }],
  table_fields: []
};

const roles: OptionProps[] = [
  { label: 'Admin', value: 'admin' },
  { label: 'User', value: 'user' }
];

export const useDomainActionProps = (_?: DomainActionProps) => {
  const history = useHistory();
  const [updatedItem, setUpdatedItem] = useState(null);
  const [isCreate, setIsCreate] = useState(true);
  const form = useForm({
    defaultValues: initialValues,
    mode: 'onBlur'
  });
  const fieldsArrayField = useFieldArray({
    control: form.control,
    name: 'fields'
  });

  const onSubmit = (values: CreateDomainForm) => {
    console.log({ values });
  };

  const fields = useMemo(
    () =>
      fieldsArrayField.fields
        ?.filter((_, i) => i !== 0)
        ?.map(field => ({
          label: field.name,
          value: field.slug
        })) || [],
    [fieldsArrayField.fields]
  );

  useEffect(() => {
    const searchObject = parseQueryToString(history.location.search);
    if (searchObject.add === 'new') {
      setIsCreate(true);
    } else if (searchObject.edit) {
      setUpdatedItem(searchObject.edit);
    }
  }, [history.location.search]);

  return {
    updatedItem,
    isCreate,
    onSubmit,
    roles,
    form,
    fieldsArrayField,
    fields
  };
};
