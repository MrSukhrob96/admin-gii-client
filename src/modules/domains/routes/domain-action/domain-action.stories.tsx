import { ComponentStory, ComponentMeta } from '@storybook/react';
import { DomainAction } from './domain-action.component';
import '@styles/index.css';

export default {
  title: 'Common/DomainAction',
  component: DomainAction,
} as ComponentMeta<typeof DomainAction>

const Template: ComponentStory<typeof DomainAction> = (args) => <DomainAction {...args} />;

export const DomainActionInitial = Template.bind({});
DomainActionInitial.args = {
}