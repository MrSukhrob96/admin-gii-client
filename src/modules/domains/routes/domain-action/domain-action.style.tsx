import styled from 'styled-components'

export const DomainActionContainer = styled.div`

`