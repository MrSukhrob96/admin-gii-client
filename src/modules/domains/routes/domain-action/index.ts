export * from './domain-action.component';
