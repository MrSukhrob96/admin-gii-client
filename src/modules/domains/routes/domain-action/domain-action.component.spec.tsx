import { customRender } from '@test-utils';
import { DomainAction } from './domain-action.component';

describe('DomainAction', () => {
  test('Should render DomainAction component', async () => {
    const { asFragment } = customRender(
      <DomainAction />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});