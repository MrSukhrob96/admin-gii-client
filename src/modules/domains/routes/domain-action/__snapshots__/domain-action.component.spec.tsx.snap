// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainAction Should render DomainAction component 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <form
      class="w-half"
    >
      <div
        class="bp4-form-group bp4-intent-success"
      >
        <label
          class="bp4-label"
          for="name-input"
        >
          Nomi 
          <span
            class="bp4-text-muted"
          >
            (*)
          </span>
        </label>
        <div
          class="bp4-form-content"
        >
          <div
            class="bp4-input-group"
          >
            <input
              class="bp4-input"
              id="name-input"
              name="name"
              placeholder="Nomini kiriting"
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        class="bp4-form-group bp4-intent-success"
      >
        <label
          class="bp4-label"
          for="slug-input"
        >
          Slag 
          <span
            class="bp4-text-muted"
          >
            (*)
          </span>
        </label>
        <div
          class="bp4-form-content"
        >
          <div
            class="bp4-input-group"
          >
            <input
              class="bp4-input"
              id="slug-input"
              name="slug"
              placeholder="Slagini kiriting"
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        class="bp4-popover-wrapper bp4-fill"
      >
        <div
          aria-haspopup="true"
          class="bp4-popover-target"
        >
          <div
            class=""
          >
            <div
              class="bp4-input bp4-tag-input bp4-fill bp4-multi-select"
            >
              <div
                class="bp4-tag-input-values"
              >
                <input
                  class="bp4-input-ghost bp4-multi-select-tag-input-input"
                  placeholder="Search..."
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        class="bp4-button bp4-intent-success"
        type="submit"
      >
        <span
          aria-hidden="true"
          class="bp4-icon bp4-icon-cloud-upload"
          icon="cloud-upload"
          tabindex="-1"
        >
          <svg
            data-icon="cloud-upload"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M8.71 7.29C8.53 7.11 8.28 7 8 7s-.53.11-.71.29l-3 3a1.003 1.003 0 001.42 1.42L7 10.41V15c0 .55.45 1 1 1s1-.45 1-1v-4.59l1.29 1.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71l-3-3zM12 4c-.03 0-.07 0-.1.01A5 5 0 002 5c0 .11.01.22.02.33a3.495 3.495 0 00.07 6.37c-.05-.23-.09-.46-.09-.7 0-.83.34-1.58.88-2.12l3-3a2.993 2.993 0 014.24 0l3 3c.54.54.88 1.29.88 2.12 0 .16-.02.32-.05.47C15.17 10.78 16 9.5 16 8c0-2.21-1.79-4-4-4z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="bp4-button-text"
        >
          Saqlash
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
